### ISSUE TITLE
Problem about code in STED_train.py / line 131

### ISSUE THREAD

**Issue by Senwang98** [olly-styles/Multiple-Object-Forecasting#14]
test_cv_preds = pd.read_csv(
            './outputs/constant_velocity/test_' + detector + '_fold_' + str(fold) + '.csv')

Hi, Olly
Now, I have saved DTP's feature. The last thing I need to do is to train STED using pre-prepared npy files.
When I tried to run STED_train.py, I found I lacked **./outputs/constant_velocity/test_' + detector + '_fold_' + str(fold) + '.csv**.
So, I have a look at downloaded output.zip which is provided by you, I found the the segmentation of the data set in the output.zip file seems to be different from your **cross-validation split.**
For example, In ./outputs/constant_velocity/test_yolo_fold1.csv which recorded PADUA and RIGA data. But, in your cross-validation split I can't find city PADUA and RIGA in 'Fold1 Test cities: ['KRAKOW' 'RIGA' 'WROCLAW']'

![image](https://user-images.githubusercontent.com/23122139/90982923-02804700-e55a-11ea-8b0b-f4e7a43615e1.png)

So, my question is that how you save outputs.zip results????
Waiting for your reply. Thx @olly-styles

**Comment by Senwang98** [comment]
I also want to  how you calculate constant velocity's upper left corner, lower right corner. DTP just prediction center point, how length and width of bounding box come from? @olly-styles

**Comment by olly-styles** [comment]
I've looked into the issue with the cross-validation split. I think there may have been a mix-up with the pre-computed feature data split. Could you try the following data split and see if it looks correct?

Fold 1
Train cities: ['BARCELONA', 'BRNO', 'DRESDEN', 'ERFURT', 'KAUNAS', 'KRAKOW', 'LEIPZIG', 'PALMA', 'POZNAN', 'TALLINN', 'VILNIUS', 'WARSAW', 'WEIMAR']

Validation cities: ['HELSINKI', 'NUREMBERG', 'PRAGUE', 'TARTU']

Test cities: ['PADUA', 'RIGA', 'VERONA', 'WROCLAW']

Fold 2
Train cities: ['BARCELONA', 'DRESDEN', 'HELSINKI', 'KRAKOW', 'NUREMBERG', 'PADUA', 'POZNAN', 'PRAGUE', 'RIGA', 'TALLINN', 'TARTU', 'VERONA', 'WARSAW', 'WROCLAW']

Validation cities: ['BRNO', 'KAUNAS', 'LEIPZIG']

Test cities: ['ERFURT', 'PALMA', 'VILNIUS', 'WEIMAR']

Fold 3
Train cities: ['BRNO', 'ERFURT', 'HELSINKI', 'KAUNAS', 'LEIPZIG', 'NUREMBERG', 'PADUA', 'PALMA', 'PRAGUE', 'RIGA', 'TARTU', 'VERONA', 'VILNIUS', 'WEIMAR', 'WROCLAW']

Validation cities: ['KRAKOW', 'POZNAN', 'TALLINN']

Test cities: ['BARCELONA', 'DRESDEN', 'WARSAW']


RE constant velocity calculation, we use the velocity of the centroid. We represent bounding boxes as the centroid, height, and width rather than each corner. This is explained in Section 5.3 of the paper.

**Comment by Senwang98** [comment]
I still feel confused. I think the prediction in steed_trian.py should be the meaning mentioned in your paper 5.3, but in the github project you said it is the upper left corner and the lower right corner(As shown below). 
![image](https://user-images.githubusercontent.com/23122139/91468987-5de96680-e8c5-11ea-82f1-eaa2dfb2b0d4.png)

I mean, how do you calculate the upper left corner and the lower right corner in the process of saving the uniform motion.

This situation does not appear in ground truth and prediction, because I do not need to manually calculate the length and width of the detection bounding box.

But when saving the information of 60 frames of uniform motion, I need to know how to deal with the length and width of the detection  bounding box. DTP helped us calculate the center point, so what principle is the calculation of length and width based on?

In other words, in Constant velocity, the center point is successfully predicted based on a uniform motion, but how to calculate the size of the detection bounding box is unknown. Obviously, according to the size of the past frame, it is unreasonable to continue to use the uniform motion strategy. If the size of the detection bounding box of DeepSort is still used, the detection bounding box of constant velocity and ground truth is the same size, which does not seem very reasonable.

**Comment by Senwang98** [comment]
So, I just want to know how to get mentioned file of constant velocity?(Not ground truth and prediction)
![image](https://user-images.githubusercontent.com/23122139/91469491-0b5c7a00-e8c6-11ea-8d54-9bd6c69bae64.png)

**Comment by olly-styles** [comment]
The constant velocity model assumes a constant object scale as well (Section 6.1, Constant Velocity & Constant Scale). In other words, the height and width of the bounding box are unchanged from the last observation. This is a simple baseline method. We did try calculating the change in height and width and linearly extrapolating these as well, though we found the results to be worse than constant height and width (again in Section 6.1, Constant Velocity & Constant Scale). 

Hope I've understood correctly and answered your question.

**Comment by Senwang98** [comment]
First of all, thank Olly for your patient explanation.

In order to avoid wasting time due to misunderstanding, I think I should reiterate my doubts.

I want to run the code after line 131 of steed_train.py, but I find that I am missing Constant Velocity.csv (preprocessed file). The ground truth and prediction files are easy to handle because they are known. When processing Constant Velocity.csv, although DTP can predict the position of the following points based on the past five frames, DTP is based on point coordinates from beginning to end, so I am curious how to obtain the length and width information, the github project interface told me , **X1_1, y1_1, x2_1, y2_1** in constant velocity.csv represent the coordinates of the upper left corner and the lower right corner. After reading your reply, I think the length and width information is taken from the previous frame of the current frame, which means that the length and width of the **t-th frame** is **equal** to the length of the **t-1th frame**. I don't know if my understanding is true or not? ?

**Comment by olly-styles** [comment]
Yes, your understanding is correct. The length and width of all predicted frames are equal to the length and width of the last observed frames. The output of the model is then the increase or decrease in height and width of the box.

Here's a concrete example: The height of the bounding box at the current timestep (t=0) is 50 pixels. The model then outputs a value of 3 for timestep 1 (t=1). The predicted height of this bounding box is therefore 53 pixels. At timestep 6 (t=6), the model outputs a value of 12. The predicted height of this bounding box is therefore 62 pixels. 

You can convert the bounding box representation from centroid, width, height to upper left, lower right using [this function](https://github.com/olly-styles/Multiple-Object-Forecasting/blob/e84cc43fd9fa4bba65844f76f6427771c690144c/utils.py#L125)
